ethstorage/es-node
# Issue164: A new node can not discover all the active peers that boot node connected

Related issues: https://github.com/ethstorage/es-node/issues/164, https://github.com/ethstorage/es-node/issues/131

if the node does not have AdvertiseIP and public IP (like run in the docker or behind a nat), we can get the public IP and the TCP port from the libp2p. host's addresses (get from the nat service or address observed by peers) and set to the node discovery. As the host's addresses would not show when it start, so we will check it every initLocalNodeAddrInterval. After the public IP and the TCP port are initialized, we will also refresh the address in case the address changes.

In this change, I fork go-ethereum to change the iptrackMinStatements (in go-ethereum\p2p\enode\localnode.go) from 10 to 4, which is used by IPTracker to predict the external endpoint, i.e. IP address and port, of the local host. (If the number of remote peers sees the same IP address and port, the IP and port will be set to localnode)

Test sample:
The following are our three nodes that run in docker, the machine behind NAT, and the machine running in Chinese. `All node` means a node in the routing table, and `peer` means peers in the sync client.
 
![b79685f722c443f27860d9963ce56d5](https://github.com/ethstorage/es-node/assets/14110556/be327c26-87b6-4d96-8d2c-2c978eeccac4)
![d55c1c6189138a9cb30192c3d6420a6](https://github.com/ethstorage/es-node/assets/14110556/79d1163a-4525-4488-b944-23365a2f671b)

How to test:
Run two es-nodes behind NAT service or docker without advertiseIP being set. and the two nodes can connect after some time.

How to verify in the real network:
If you are running a private node, you can compare its peer count with a public node's peer count, if the values are similar, it should work fine. if the peer count of the private node is much smaller than the public node, then we need more investigation.
